ui/core/worker_pool.ts

```diff
@@ -81,12 +83,14 @@ export class WorkerPool {
 		try {
 			worker.addSimTaskRunning(id, request.simOptions ? request.simOptions.iterations * request.statsToWeigh.length : 30000);
 
-			// Add handler for the progress events
-			worker.addPromiseFunc(this.getProgressName(id), this.newProgressHandler(id, worker, onProgress), noop);
-
 			// Now start the async sim
-			const resultData = await worker.doApiCall(SimRequest.statWeightsAsync, StatWeightsRequest.toBinary(request), id);
-			const result = ProgressMetrics.fromBinary(resultData);
+			worker.doApiCall(SimRequest.statWeightsAsync, StatWeightsRequest.toBinary(request), id);
+			// Wait for final data
+			const result: ProgressMetrics = await new Promise(resolve => {
```

Comment:
Since this fn is used multiple times, can we perhaps make this a separate function? 
```
this.waitForFinalData(id, worker, onProgress)

async waitForFinalData(id, worker, onProgress): ProgressMetrics {
  return new Promise(resolve => {
		// Add handler for the progress events
		worker.addPromiseFunc(this.getProgressName(id), this.newProgressHandler(id, worker, onProgress, pm => resolve(pm)), noop);
  });
}
```